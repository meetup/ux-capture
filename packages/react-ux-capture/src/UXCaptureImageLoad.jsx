// @flow
import * as React from 'react';

type Props = React$ElementConfig<typeof HTMLImageElement> & { mark: string };

export const getOnLoadJS = (mark: string): string => {
	const onload = `
		if (window.UXCapture && !window.UXCapture['${mark}-LOADED']) {
			window.UXCapture.mark('${mark}');
			window.UXCapture['${mark}-LOADED'] = true;
		}
	`;

	// Replace newlines and tabs with space characters
	return onload.replace(/[\n\t]+/g, ' ');
};

export const getPropsAsHTMLAttrs = (
	props: React$ElementConfig<typeof HTMLImageElement>
): string => {
	return Object.keys(props)
		.map(prop => {
			let attribute = prop;

			// convert JSX attribute names to HTML syntax
			// Any other special props needed? Add here.
			// @see: https://reactjs.org/docs/dom-elements.html
			if (prop === 'className') {
				attribute = 'class';
			}
			if (prop === 'htmlFor') {
				attribute = 'for';
			} else {
				// handles camelCased attributes like `onLoad`
				attribute = attribute.toLowerCase();
			}

			return `${attribute}="${props[prop]}"`;
		})
		.join(' ');
};

/**
 * Creates an image tag with provide props
 * and a UXCapture `onLoad` handler
 *
 * @see example https://github.com/meetup/ux-capture#image-elements
 */
const UXCaptureImageLoad = (props: Props) => {
	const { mark, src, ...other } = props;

	const onload = getOnLoadJS(mark);
	const otherImgAttrs = getPropsAsHTMLAttrs(other);

	return (
		<div
			dangerouslySetInnerHTML={{
				__html: `
				<img id="ux-capture-${mark}" src="${src}" onload="${onload}" ${otherImgAttrs} />
			`,
			}}
		/>
	);
};

export default UXCaptureImageLoad;
