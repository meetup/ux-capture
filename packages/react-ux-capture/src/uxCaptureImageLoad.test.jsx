import React from 'react';
import { shallow } from 'enzyme';
import UXCaptureImageLoad, {
	getOnLoadJS,
	getPropsAsHTMLAttrs,
} from './UXCaptureImageLoad';

const MOCK_PROPS = {
	mark: 'ux-image-load-logo',
	src: '/foo.png',
	className: 'someClass',
	alt: 'altText',
	height: '50',
};

const renderComponent = (props = MOCK_PROPS) =>
	shallow(<UXCaptureImageLoad {...props} />);

describe('getOnLoadJS', () => {
	it('renders the correct javascript as a string', () => {
		expect(getOnLoadJS(MOCK_PROPS.mark)).toMatchSnapshot();
	});
});

describe('getPropsAsHTMLAttrs', () => {
	it('renders the correct javascript as a string', () => {
		const imageProps = {
			className: MOCK_PROPS.className,
			alt: MOCK_PROPS.alt,
			height: MOCK_PROPS.height,
		};

		expect(getPropsAsHTMLAttrs(imageProps)).toMatchSnapshot();
	});
});

describe('UXCaptureImageLoad', () => {
	it('renders component markup', () => {
		global.window = { UX: { mark: jest.fn() } };
		expect(renderComponent()).toMatchSnapshot();
		delete global.window;
	});
});
