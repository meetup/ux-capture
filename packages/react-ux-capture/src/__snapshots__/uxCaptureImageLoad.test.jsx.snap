// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UXCaptureImageLoad renders component markup 1`] = `
<span
  dangerouslySetInnerHTML={
    Object {
      "__html": "
				<img src=\\"/foo.png\\" onload=\\" if (window.UXCapture && !window.UXCapture['ux-image-load-logo-LOADED']) { window.UXCapture.mark('ux-image-load-logo'); window.UXCapture['ux-image-load-logo-LOADED'] = true; } \\"  class=\\"someClass\\" alt=\\"altText\\" height=\\"50\\" />
			",
    }
  }
/>
`;

exports[`getOnLoadJS renders the correct javascript as a string 1`] = `" if (window.UXCapture && !window.UXCapture['ux-image-load-logo-LOADED']) { window.UXCapture.mark('ux-image-load-logo'); window.UXCapture['ux-image-load-logo-LOADED'] = true; } "`;

exports[`getPropsAsHTMLAttrs renders the correct javascript as a string 1`] = `"class=\\"someClass\\" alt=\\"altText\\" height=\\"50\\""`;
